src/main/java/hudson/remoting/Engine.java

```diff
@@ -678,6 +643,23 @@ public void afterChannel(@Nonnull JnlpConnectionState event) {
         }
     }
 
+    private JnlpEndpointResolver createEndpointResolver(List<String> jenkinsUrls) {
```

Comment:
Create either a traditional resolver that downloads the JNLP file and extracts the information from it or a new one that just builds the necessary endpoint information from parameters.